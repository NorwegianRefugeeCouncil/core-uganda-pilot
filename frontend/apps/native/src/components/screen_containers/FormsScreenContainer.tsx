import { FormDefinition } from 'core-api-client';
import React from 'react';

import { FormsScreenContainerProps } from '../../types/screens';
import useApiClient from '../../utils/clients';
import { FormsScreen } from '../screens/FormsScreen';

export const FormsScreenContainer = ({
  navigation,
  route,
}: FormsScreenContainerProps) => {
  const [forms, setForms] = React.useState<FormDefinition[]>();
  const [isLoading, setIsLoading] = React.useState(true);
  const apiClient = useApiClient();

  React.useEffect(() => {
    apiClient.Form.list({}).then((data) => {
      setIsLoading(false);
      setForms(data.response?.items);
    });
  }, []);
  return (
    <FormsScreen isLoading={isLoading} forms={forms} navigation={navigation} />
  );
};
