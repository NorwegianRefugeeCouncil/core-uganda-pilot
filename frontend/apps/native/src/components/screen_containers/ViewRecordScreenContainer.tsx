import React from "react";
import ViewRecordScreen, {ViewRecordScreenProps} from "../screens/ViewRecordScreen";
import {ViewRecordScreenContainerProps} from "../../types/screens";
import client from "../../utils/clients";
import {useForm} from "react-hook-form";
import {FormDefinition} from "core-api-client";

export const ViewRecordScreenContainer = ({route, state}: ViewRecordScreenContainerProps) => {
    const {formId, recordId} = route.params;

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();

    const {control, reset} = useForm();

    React.useEffect(() => {
        const getForm = async () => {
            try {
                const {response} = await client().getForm({id: formId});
                setForm(response);
            } catch (err) {
                console.error(err);
            }
        };
        getForm();
    }, [formId]);

    React.useEffect(() => {
        if (form) {
            reset(state.formsById[formId].recordsById[recordId].values);
            setIsLoading(false);
        }
    }, [form]);

    const viewRecordScreenProps: ViewRecordScreenProps = {
        isLoading,
        form,
        control,
    };
    return <ViewRecordScreen {...viewRecordScreenProps} />;
};
