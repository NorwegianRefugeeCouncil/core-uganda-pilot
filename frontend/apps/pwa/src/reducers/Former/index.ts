import { createSlice } from '@reduxjs/toolkit';

import { adapter } from './former.adapter';
import { extraReducers, reducers } from './former.reducers';

export default createSlice({
  name: 'recorder',
  initialState: {
    ...adapter.getInitialState(),
    selectedFormId: '',
    selectedFieldId: undefined as string | undefined,
    selectedDatabaseId: undefined as string | undefined,
    selectedFolderId: undefined as string | undefined,
    savePending: false,
    saveSuccess: false,
    saveId: undefined as string | undefined,
    saveError: undefined,
  },
  reducers,
  extraReducers,
});
