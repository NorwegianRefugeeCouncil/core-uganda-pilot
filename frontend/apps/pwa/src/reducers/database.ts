import { createAsyncThunk, createEntityAdapter, createSlice, EntityState } from '@reduxjs/toolkit';
import { Database, DatabaseList, DatabaseListRequest, Response } from 'core-api-client';

import { RootState } from '../app/store';
import client from '../app/client';

const adapter = createEntityAdapter<Database>({
  // Assume IDs are stored in a field other than `book.id`
  selectId: (db) => db.id,
  // Keep the "all IDs" array sorted based on book titles
  sortComparer: (a, b) => a.name.localeCompare(b.name),
});

export const fetchDatabases = createAsyncThunk<Response<DatabaseListRequest, DatabaseList>>(
  'databases/fetch',
  async (_, thunkAPI) => {
    try {
      const response = await client.listDatabases({});
      if (response.success) {
        return response;
      }
      return thunkAPI.rejectWithValue(response);
    } catch (err) {
      return thunkAPI.rejectWithValue(err);
    }
  },
);

export interface DatabaseState extends EntityState<Database> {
  fetchPending: boolean;
  fetchError: any;
  fetchSuccess: boolean;
}

export const databasesSlice = createSlice({
  name: 'databases',
  initialState: {
    ...adapter.getInitialState(),
    fetchPending: false,
    fetchError: undefined as any,
    fetchSuccess: true,
  },
  reducers: {
    addOne: adapter.addOne,
    addMany: adapter.addMany,
    removeAll: adapter.removeAll,
    removeMany: adapter.removeMany,
    removeOne: adapter.removeOne,
    updateMany: adapter.updateMany,
    updateOne: adapter.updateOne,
    upsertOne: adapter.upsertOne,
    upsertMany: adapter.upsertMany,
    setOne: adapter.setOne,
    setMany: adapter.setMany,
    setAll: adapter.setAll,
  },
  extraReducers: (builder) => {
    builder.addCase(fetchDatabases.pending, (state, action) => {
      state.fetchSuccess = false;
      state.fetchPending = true;
    });
    builder.addCase(fetchDatabases.rejected, (state, action) => {
      state.fetchSuccess = false;
      state.fetchPending = false;
      state.fetchError = action.payload;
    });
    builder.addCase(fetchDatabases.fulfilled, (state, action) => {
      state.fetchSuccess = true;
      state.fetchPending = false;
      state.fetchError = undefined;
      if (action.payload.response?.items) {
        adapter.setAll(state, action.payload.response.items);
      }
    });
  },
});

export const databaseActions = databasesSlice.actions;
const selectors = adapter.getSelectors();

export const databaseSelectors = {
  ...selectors,
};

export const databaseGlobalSelectors = adapter.getSelectors<RootState>((state) => state.databases);

export default databasesSlice.reducer;
