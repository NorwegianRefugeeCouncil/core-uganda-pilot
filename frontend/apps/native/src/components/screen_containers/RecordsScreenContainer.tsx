import React from 'react';

import { RECORD_ACTIONS } from '../../reducers/recordsReducers';
import { RecordsScreenContainerProps } from '../../types/screens';
import RecordsScreen from '../screens/RecordsScreen';
import client from '../../utils/clients';

export const RecordsScreenContainer = ({ navigation, route, state, dispatch }: RecordsScreenContainerProps) => {
  const { formId, databaseId } = route.params;
  const [isLoading, setIsLoading] = React.useState(true);

  React.useEffect(() => {
    if (formId && databaseId) {
      client()
        .listRecords({ formId, databaseId })
        .then((data) => {
          dispatch({
            type: RECORD_ACTIONS.GET_RECORDS,
            payload: {
              formId,
              records: data.response?.items,
            },
          });
        })
        .catch(console.error)
        .finally(() => setIsLoading(false));
    }
  }, [client, formId, databaseId]);

  const recordsScreenProps = { isLoading, state, navigation, route };
  return <RecordsScreen {...recordsScreenProps} />;
};
