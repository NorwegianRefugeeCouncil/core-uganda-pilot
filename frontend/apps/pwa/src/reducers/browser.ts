import { Folder, FormDefinition } from 'core-api-client';

import { RootState } from '../app/store';

import { selectByFolderOrDBId } from './form';
import { selectByParentId, selectDatabaseRootFolders } from './folder';
import { databaseGlobalSelectors } from './database';

export const selectChildFolders =
  (dbOrFolderId?: string) =>
  (state: RootState): Folder[] => {
    if (!dbOrFolderId) {
      return [];
    }
    const db = databaseGlobalSelectors.selectById(state, dbOrFolderId);
    if (db) {
      return selectDatabaseRootFolders(state, db.id);
    }
    return selectByParentId(state, dbOrFolderId);
  };

export const selectChildForms =
  (ids: { dbId?: string; folderId?: string }) =>
  (state: RootState): FormDefinition[] => {
    return selectByFolderOrDBId(state, ids);
  };
