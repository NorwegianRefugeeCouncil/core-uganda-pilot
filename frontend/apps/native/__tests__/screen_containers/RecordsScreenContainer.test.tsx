import React from 'react';
import { render, waitFor } from '@testing-library/react-native';

import { RecordsScreenContainer } from '../../src/components/screen_containers/RecordsScreenContainer';
import { Record } from '../../../client/lib/esm';
import { RECORD_ACTIONS } from '../../src/reducers/recordsReducers';

const mockFormId = 'a form id';
const mockDatabaseId = 'a database id';
const mockRoute = {
  params: {
    formId: mockFormId,
    databaseId: mockDatabaseId,
  },
};
const mockNavigation = {};
const mockState: any = {
  formsById: {},
};

const mockDispatch = jest.fn();

const mockRemoteRecord: Record = {
  id: '',
  databaseId: mockDatabaseId,
  formId: mockFormId,
  values: {},
  ownerId: '',
};

const mockRemoteRecords = Array(10)
  .fill(mockRemoteRecord)
  .map((r, i) => ({ ...r, id: i.toString() }));

const mockLocalRecord = '';
const mockLocalRecords = Array(10)
  .fill(mockLocalRecord)
  .map((_, i) => `local record #${i}`);

const mockRemoteData = {
  response: {
    items: mockRemoteRecords,
  },
};

mockState.formsById[mockFormId] = {
  records: mockRemoteRecords,
  localRecords: mockLocalRecords,
};

const mockProps: any = {
  route: mockRoute,
  navigation: mockNavigation,
  state: mockState,
  dispatch: mockDispatch,
};

const mockListRecords = jest.fn().mockResolvedValue(mockRemoteData);
jest.mock('../../src/utils/clients', () => () => ({
  listRecords: (object: any) => mockListRecords(object),
}));

describe(RecordsScreenContainer.name, () => {
  afterEach(() => {
    mockListRecords.mockClear();
  });

  test('renders correctly', async () => {
    const { toJSON } = await waitFor(() => render(<RecordsScreenContainer {...mockProps} />));
    expect(toJSON()).toMatchSnapshot();
  });

  test('fetches records and dispatches the appropriate action', async () => {
    await waitFor(() => render(<RecordsScreenContainer {...mockProps} />));

    const expectedListRecordsArg = {
      formId: mockFormId,
      databaseId: mockDatabaseId,
    };
    expect(mockListRecords).toHaveBeenCalledWith(expectedListRecordsArg);

    const expectedDispatchAction = {
      type: RECORD_ACTIONS.GET_RECORDS,
      payload: {
        formId: mockFormId,
        records: mockRemoteRecords,
      },
    };
    expect(mockDispatch).toHaveBeenCalledWith(expectedDispatchAction);
  });

  test('does not attempt to fetch if missing formId on route', async () => {
    const incompleteRoute = {
      ...mockRoute,
      params: {
        ...mockRoute.params,
        formId: null,
      },
    };
    await waitFor(() => render(<RecordsScreenContainer {...mockProps} route={incompleteRoute} />));

    expect(mockListRecords).not.toHaveBeenCalled();
  });

  test('does not attempt to fetch if missing databaseId on route', async () => {
    const incompleteRoute = {
      ...mockRoute,
      params: {
        ...mockRoute.params,
        databaseId: null,
      },
    };
    await waitFor(() => render(<RecordsScreenContainer {...mockProps} route={incompleteRoute} />));

    expect(mockListRecords).not.toHaveBeenCalled();
  });
});
