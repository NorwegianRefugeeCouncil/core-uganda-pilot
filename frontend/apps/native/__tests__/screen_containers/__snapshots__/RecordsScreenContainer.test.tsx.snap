// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordsScreenContainer renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexGrow": 1,
      },
      Object {
        "alignItems": "flex-start",
        "flexGrow": 1,
        "justifyContent": "flex-start",
        "padding": 18,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Object {
          "textAlign": "left",
        },
        Array [
          Object {
            "textAlign": "left",
          },
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "500",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 20,
              "letterSpacing": 0.15,
              "lineHeight": 30,
              "marginVertical": 2,
            },
            undefined,
          ],
        ],
      ]
    }
  >
    Records
  </Text>
  <View>
    <RCTScrollView
      data={
        Array [
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "0",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "1",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "2",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "3",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "4",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "5",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "6",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "7",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "8",
            "ownerId": "",
            "values": Object {},
          },
          Object {
            "databaseId": "a database id",
            "formId": "a form id",
            "id": "9",
            "ownerId": "",
            "values": Object {},
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={true}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "width": "100%",
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  0
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  1
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  2
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  3
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  4
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  5
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  6
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  7
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  8
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="view record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="remote record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  9
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <RCTScrollView
      data={
        Array [
          "local record #0",
          "local record #1",
          "local record #2",
          "local record #3",
          "local record #4",
          "local record #5",
          "local record #6",
          "local record #7",
          "local record #8",
          "local record #9",
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={true}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "width": "100%",
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #0
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #1
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #2
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #3
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #4
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #5
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #6
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #7
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #8
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="add record"
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="local record"
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Object {
                    "justifyContent": "center",
                    "paddingRight": 12,
                  }
                }
              >
                <Text>
                  local record #9
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    pointerEvents="auto"
    style={
      Object {
        "backgroundColor": "#03dac4",
        "borderRadius": 28,
        "bottom": 28,
        "elevation": 6,
        "opacity": 1,
        "position": "absolute",
        "right": 18,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 6,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 28,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "height": 56,
              "width": 56,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "height": 24,
                "width": 24,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "material-community",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰐕
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
